### PR TITLE
Fix build errors and warnings in NetBSD.

### PR DESCRIPTION
**Pull request by fraggerfox**
- Fix build in NetBSD from the renaming done in b74673fe.
- Simplify the `ProcessTable_goThroughEntries()` and also apply the fixes for the compiler warning.
  ```
  netbsd/NetBSDProcessTable.c: In function 'NetBSDProcessTable_scanProcs':
  netbsd/NetBSDProcessTable.c:185:53: warning: comparison is always true due to limited range of data type [-Wtype-limits]
           const char* name = ((int64_t)kproc->p_tdev != (int64_t)KERN_PROC_TTY_NODEV) ? devname(kproc->p_tdev, S_IFCHR): NULL;
  ```
  The fix came in as part of the discussion [here](https://github.com/htop-dev/htop/pull/1360#discussion_r1439257379).

### CODE REVIEW

**Comment by Explorer09** [comment]
Do you have a build log, i.e. a list of compiler warnings and errors of your build from the original code? That would save us time reviewing the code.

**Comment by fraggerfox** [comment]
> Do you have a build log, i.e. a list of compiler warnings and errors of your build from the original code? That would save us time reviewing the code.

Yes, this makes sense, should have added it when I was making PR.

**Comment by fraggerfox** [comment]
```
netbsd/NetBSDProcess.c: In function 'NetBSDProcess_rowWriteField':
netbsd/NetBSDProcess.c:240:26: error: incompatible type for argument 1 of 'Process_writeField'
       Process_writeField(np->super, str, field);
                          ^~
In file included from ./netbsd/NetBSDProcess.h:17:0,
                 from netbsd/NetBSDProcess.c:13:
./Process.h:219:6: note: expected 'const Process * {aka const struct Process_ *}' but argument is of type 'Process {aka const struct Process_}'
 void Process_writeField(const Process* row, RichString* str, ProcessField field);
      ^~~~~~~~~~~~~~~~~~
*** [netbsd/NetBSDProcess.o] Error code 1

make[1]: stopped in /usr/pkgsrc-current/wip/htop-git/work/htop-c3b4a770c02c28662980014b5a893e01451ec2b2
--- netbsd/NetBSDMachine.o ---
netbsd/NetBSDMachine.c: In function 'NetBSDMachine_updateCPUcount':
netbsd/NetBSDMachine.c:84:7: error: 'opl' undeclared (first use in this function); did you mean 'open'?
       opl->cpuData = xReallocArray(this->cpuData, value + 1, sizeof(CPUData));
       ^~~
       open
netbsd/NetBSDMachine.c:84:7: note: each undeclared identifier is reported only once for each function it appears in
netbsd/NetBSDMachine.c: In function 'Machine_new':
netbsd/NetBSDMachine.c:130:11: warning: return from incompatible pointer type [-Wincompatible-pointer-types]
    return this;
           ^~~~
netbsd/NetBSDMachine.c: In function 'Machine_scan':
netbsd/NetBSDMachine.c:271:4: warning: implicit declaration of function 'NetBSDProcessTable_scanMemoryInfo'; did you mean 'NetBSDMachine_scanMemoryInfo'? [-Wimplicit-function-declaration]
    NetBSDProcessTable_scanMemoryInfo(this);
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    NetBSDMachine_scanMemoryInfo
netbsd/NetBSDMachine.c:272:4: warning: implicit declaration of function 'NetBSDProcessTable_scanCPUTime'; did you mean 'NetBSDMachine_scanCPUTime'? [-Wimplicit-function-declaration]
    NetBSDProcessTable_scanCPUTime(this);
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    NetBSDMachine_scanCPUTime
netbsd/NetBSDMachine.c:275:7: warning: implicit declaration of function 'NetBSDProcessTable_scanCPUFrequency'; did you mean 'NetBSDMachine_scanCPUFrequency'? [-Wimplicit-function-declaration]
       NetBSDProcessTable_scanCPUFrequency(npl);
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
       NetBSDMachine_scanCPUFrequency
netbsd/NetBSDMachine.c:275:43: error: 'npl' undeclared (first use in this function); did you mean 'nl'?
       NetBSDProcessTable_scanCPUFrequency(npl);
                                           ^~~
                                           nl
At top level:
netbsd/NetBSDMachine.c:222:13: warning: 'NetBSDMachine_scanCPUFrequency' defined but not used [-Wunused-function]
 static void NetBSDMachine_scanCPUFrequency(NetBSDMachine* this) {
             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
netbsd/NetBSDMachine.c:197:13: warning: 'NetBSDMachine_scanCPUTime' defined but not used [-Wunused-function]
 static void NetBSDMachine_scanCPUTime(NetBSDMachine* this) {
             ^~~~~~~~~~~~~~~~~~~~~~~~~
netbsd/NetBSDMachine.c:145:13: warning: 'NetBSDMachine_scanMemoryInfo' defined but not used [-Wunused-function]
 static void NetBSDMachine_scanMemoryInfo(NetBSDMachine* this) {
             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
*** [netbsd/NetBSDMachine.o] Error code 1
```
Relevant section of the build log showing the failure.

File: netbsd/NetBSDProcessTable.c
Comment (fraggerfox) [inline netbsd/NetBSDProcessTable.c:185]:
```c
(dev_t)kproc->p_tdev != KERN_PROC_TTY_NODEV)
```

There was a compiler warning stating this expression always evaluates to true. so removed it. The new build seems to be working fine after this removal.

Comment (Explorer09) [reply netbsd/NetBSDProcessTable.c:185]: I dug the NetBSD header files, and this looks like a bug in the conditional. It probably meant to work like this:

```c
         const char* name = ((int64_t)kproc->p_tdev != (int64_t)KERN_PROC_TTY_NODEV) ? devname(kproc->p_tdev, S_IFCHR) : NULL;
```

Better make this change a separate commit as this one is a bug fix rather than a build error fix.

Comment (Explorer09) [reply netbsd/NetBSDProcessTable.c:185]: All relevant definitions in NetBSD headers:

```c
// NetBSD <sys/types.h>
typedef uint64_t dev_t;
// NetBSD <sys/param.h>
#define NODEV (dev_t)(-1)
// NetBSD <sys/sysctl.h>
#define KERN_PROC_TTY_NODEV NODEV	
struct kinfo_proc2 {
...
uint32_t p_tdev;
...
};
```

Comment (fraggerfox) [reply netbsd/NetBSDProcessTable.c]: I had attempted this change when I was fixing the build and it still did not resolve the issue

```
netbsd/NetBSDProcessTable.c: In function 'NetBSDProcessTable_scanProcs':
netbsd/NetBSDProcessTable.c:185:53: warning: comparison is always true due to limited range of data type [-Wtype-limits]
          const char* name = ((int64_t)kproc->p_tdev != (int64_t)KERN_PROC_TTY_NODEV) ? devname(kproc->p_tdev, S_IFCHR): NULL;
```

I have reverted the change for now, if it causes trouble in the future will revisit it.

Comment (Explorer09) [reply netbsd/NetBSDProcessTable.c]: @fraggerfox I made a typo. `uint32_t` cast to `int64_t` always results in a positive value.

Now I'm troubled. While the downcast like the following code would work, I really don't want to depend on the type of 
`(struct kinfo_proc2).p_tdev` for portability reason.

```c
         const char* name = (kproc->p_tdev != (uint32_t)KERN_PROC_TTY_NODEV) ? devname(kproc->p_tdev, S_IFCHR) : NULL;
```

Since I am not the author or maintainer of htop NetBSD port. I would stop here and let someone who is more familiar with NetBSD code to look at this.

Comment (fraggerfox) [reply netbsd/NetBSDProcessTable.c]: @Explorer09 : I agree with you, I think I need to look a bit deeper into the meaning of this check and why I put there in the first place. For now I see no harm in leaving it as it is.

P.S: I authored the [NetBSD](https://github.com/htop-dev/htop/pull/603) implementation a while back and I currently maintain it.

Comment (Explorer09) [reply netbsd/NetBSDProcessTable.c]: @fraggerfox 
I finally figured out what to do. :)
The purpose of the code is to not allocate the name of the controlling TTY for processes that don't have one. (NetBSD source code shows that they do use `NODEV` placeholder value for `kproc->p_tdev`.) Although it might be safe to call [devname(3)](https://man.netbsd.org/devname.3) for a "no TTY device" placeholder. The `devname` function would return a placeholder string "??" which would not be ideal for displaying in htop. Hence the filter.

As `KERN_PROC_TTY_NODEV` is a negative constant. Compare with `~(KERN_PROC_TTY_NODEV)` would get rid of the signed and unsigned problem.

```c
         // KERN_PROC_TTY_NODEV is a negative constant but the type of 
         // kproc->p_tdev may be unsigned. 
         const char* name = ((dev_t)~kproc->p_tdev != (dev_t)~(KERN_PROC_TTY_NODEV)) ? devname(kproc->p_tdev, S_IFCHR) : NULL;
```

Comment (fraggerfox) [reply netbsd/NetBSDProcessTable.c]: @Explorer09 : Thank you for doing that investigation, shall I include your suggestion into this PR? (Now I remember about the `??` characters when I was doing the port originally)

I did a build locally and it seems to have worked!

Comment (Explorer09) [reply netbsd/NetBSDProcessTable.c]: @fraggerfox You can include it, but as a separate commit. This way if another BSD port of htop has the problem people can adapt the patch easily.

Comment (fraggerfox) [reply netbsd/NetBSDProcessTable.c]: Agreed, I shall do a follow up PR once this is merged.

File: netbsd/NetBSDProcessTable.c
Comment (BenBE) [inline netbsd/NetBSDProcessTable.c:154]: As `this->super` is handed to `ProcessTable_getProcess` below it might be worth splitting this assignment and extract `super` into it's own variable as is already done with most other platform implementations of `*ProcessTable_scanProcs`.

**Review by BenBE — APPROVED** [review]
Just a very small idea. Apart from this the changes look reasonable.

IIRC there was a similar PR done for FreeBSD/OpenBSD already. Maybe compare there too for changes.

File: netbsd/NetBSDProcessTable.c
Comment (fraggerfox) [reply netbsd/NetBSDProcessTable.c:154]: I had a look at this and simplified it based on what I saw in the `darwin` port.

- https://github.com/htop-dev/htop/blob/main/darwin/DarwinProcessTable.c#L72

I created a follow up PR against my branch to show the changes.

@BenBE : Would you like me to commit that also into this PR and merge both the commits in this single PR?

Comment (BenBE) [reply netbsd/NetBSDProcessTable.c:154]: Go ahead and cherry-pick. Changes look easy enough.

Comment (fraggerfox) [reply netbsd/NetBSDProcessTable.c]: Done.

**Review by BenBE — APPROVED** [review]
LGTM.

**Pull request merged**
